api/classes/prerequisites.js

```diff
@@ -0,0 +1,137 @@
+ * to fetch all matching records of the passed `model`,
+ * using data from the route parameters or the request payload
+ * to build the query.
+ *
+ * @param {Object} model - Bookshelf model for querying
+ * @param {String} mode - 'param' or 'payload' (if omitted, returns all records)
+ * @param {String} requestKey - key of the param/payload
+ * @param {String} databaseKey - key of the database model
+ *
+ * @return {Promise}
+ */
+prerequisites.confirmRecordExists = function(model, mode, requestKey, databaseKey) {
+  databaseKey = databaseKey || requestKey;
+
+  return {
+    assign: 'records',
+    method: function(req, reply) {
+      var options = {};
+      if (requestKey) {
+        options.where = {};
+        options.where[databaseKey] = mode === 'param' ? req.params[requestKey] : req.payload[requestKey];
+      }
+
+      var result = model.query(options)
+        .fetchAll()
+        .then(function(records) {
+          if (!records) { throw Boom.notFound(); }
+
+          return records;
+        })
+        .catch(errors.generateErrorResponse);
+
+      return reply(result);
+    }
+  };
+};
+
+/**
+ * validateOwnership()
+ *
+ * Ensures the authenticated user is the owner of the
+ * resource being manipulated or requested.
+ *
+ * @return {Promise}
+ */
+prerequisites.validateOwnership = function() {
+  return {
+    method: function validateOwnership(req, reply) {
+      var result = Users.query({
+        where: {
+          username: req.auth.credentials.username
+        }
+      }).fetch()
+      .then(function(record) {
+        if (!record) {
+          // This case means our auth logic failed unexpectedly
+          throw Boom.badImplementation("User doesn't exist!");
+        }
+
+        // We have to check if it's a collection,
+        // and only need to examine one record
+        var resource = req.pre.record.length ? req.pre.record[0] : req.pre.record;
+        if (resource.user_id !== record.id) {
+          throw Boom.unauthorized();
+        }
+      })
+      .catch(errors.generateErrorResponse);
+
+      return reply(result);
+    }
+  };
+};
+
+/**
+ * validateCreationPermission([foreignKey, model])
+ *
+ * Ensures the authenticated user is the owner of the
+ * resource being created.
+ *
+ * @param {Object} foreignKey - Foreign key of an existing resource
+ * @param {Object} model - Bookshelf model for querying
+ *
+ * @return {Promise}
+ */
+prerequisites.validateCreationPermission = function(foreignKey, model) {
+  return {
+    method: function validateCreationPermission(req, reply) {
+      var result = Users.query({
+        where: {
+          username: req.auth.credentials.username
+        }
+      }).fetch()
+      .then(function(userRecord) {
+        if (!userRecord) {
+          // This case means our auth logic failed unexpectedly
+          throw Boom.badImplementation("User doesn't exist!");
+        }
+
+        // There's a direct reference to the user_id in the payload
```

Comment:
This is confusing. Maybe, `Check to see if there's a direct reference to `user_id` in the payload`